probe-rs/src/architecture/arm/component/mod.rs

```diff
@@ -90,126 +110,146 @@ fn find_component(
-fn setup_swv_vendor(
+pub(crate) fn read_trace_memory(
     interface: &mut Box<dyn ArmProbeInterface>,
     components: &[CoresightComponent],
-    config: &SwoConfig,
-) -> Result<(), Error> {
-    if components.is_empty() {
-        return Err(Error::architecture_specific(RomTableError::NoComponents));
+) -> Result<Vec<u8>, Error> {
+    let mut tmc =
+        TraceMemoryController::new(interface, find_component(components, PeripheralType::Tmc)?);
+
+    // TODO: In the future, it may be possible to dynamically read from trace memory
+    // without waiting for the FIFO to fill first.
+    let fifo_size = tmc.fifo_size()?;
+    while !tmc.full()? {
+        log::info!("TMC FIFO Level: {} of {fifo_size}", tmc.fill_level()?)
     }
 
-    for component in components.iter() {
-        let mut memory = interface.memory_interface(component.ap)?;
-
-        match component.component.id().peripheral_id().jep106() {
-            Some(id) if id == jep106::JEP106Code::new(0x00, 0x20) => {
-                // STMicroelectronics:
-                // STM32 parts need TRACE_IOEN set to 1 and TRACE_MODE set to 00.
-                log::debug!("STMicroelectronics part detected, configuring DBGMCU");
-                const DBGMCU: u64 = 0xE004_2004;
-                let mut dbgmcu = memory.read_word_32(DBGMCU)?;
-                dbgmcu |= 1 << 5;
-                dbgmcu &= !(0b00 << 6);
-                return memory.write_word_32(DBGMCU, dbgmcu);
-            }
-            Some(id) if id == jep106::JEP106Code::new(0x02, 0x44) => {
-                // Nordic VLSI ASA
-                log::debug!("Nordic part detected, configuring CLOCK TRACECONFIG");
-                const CLOCK_TRACECONFIG: u64 = 0x4000_055C;
-                let mut traceconfig: u32 = 0;
-                traceconfig |= match config.tpiu_clk() {
-                    4_000_000 => 3,
-                    8_000_000 => 2,
-                    16_000_000 => 1,
-                    32_000_000 => 0,
-                    tpiu_clk => {
-                        let e = ComponentError::NordicUnsupportedTPUICLKValue(tpiu_clk);
-                        log::error!("{:?}", e);
-                        return Err(Error::architecture_specific(e));
-                    }
-                };
-                traceconfig |= 1 << 16; // tracemux : serial = 1
-                return memory.write_word_32(CLOCK_TRACECONFIG, traceconfig);
-            }
-            _ => {
-                continue;
+    // This sequence is taken from "CoreSight Trace memory Controller Technical Reference Manual"
+    // Section 2.2.2 "Software FIFO Mode". Without following this procedure, the trace data does
+    // not properly stop even after disabling capture.
+    tmc.stop_on_flush(true)?;
+    tmc.manual_flush()?;
+
+    // Read all of the data from the ETM into a vector for further processing.
+    let mut etf_trace = std::io::Cursor::new(vec![0; tmc.fifo_size()? as usize + 128]);
```

Comment:
Why make it 128 bytes bigger? How do we track actual size?

Reply:
Indeed this is poor practice. I had copied some code over from another application where std::io::Cursor was being used differently, but here all we really need is a Vec. I've replaced this whole section and simplified it to use vectors and slice chunking instead.